### PR TITLE
Large refactor for major version bump

### PR DESCRIPTION
**Pull request by mjp41**
This contains the accumulated changes to refactor snmalloc mostly targetted at security and hardening, but the redesign has effected the whole code base.

The largest changes are 
* A new pagemap that contains all the out of band meta-data
* All in band meta-data can now be protected with a novel free list protection (enabled with SNMALLOC_CHECK_CLIENT)
* Randomisation can be enables in the checked mode (enabled with SNMALLOC_CHECK_CLIENT)
* A new scheme for periodically returning memory to the OS.
* Support for CHERI 
* Cleaner factoring between front-end and back-end to enable alternative scenarios such as compartmentalisation.

This is a large refactor, and I would like to get some external perf eyes on this where possible, so I can address the issues before any new release.  So if anyone is able to run this branch through their benchmarking it would be really useful (@licenser, @SchrodingerZhu, @darach, @ryancinsight, @houqp, @Dandandan, @fafhrd91)

### CODE REVIEW

**Comment by Licenser** [comment]
I can't claim to understand all of the code :D but this looks amazing!

File: src/backend/pagemap.h [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+      size_t request_size = REQUIRED_SIZE + additional_size;
+#else
+      size_t request_size = REQUIRED_SIZE;
+#endif
+
+      auto new_body_untyped = PAL::reserve(request_size);
+
+      if (new_body_untyped == nullptr)
+      {
+        PAL::error("Failed to initialisation snmalloc.");
```
Comment (achamayou) [inline src/backend/pagemap.h:172]:
```suggestion
        PAL::error("Failed to initialise snmalloc.");
```

File: src/pal/pal_open_enclave.h [hunk 2]
```diff
@@ -1,66 +1,37 @@
...
-    static constexpr uint64_t pal_features = Entropy;
-
-    static constexpr size_t page_size = Aal::smallest_page_size;
+    static void print_stack_trace() {}
 
     [[noreturn]] static void error(const char* const str)
     {
       UNUSED(str);
       oe_abort();
     }
```
Comment (achamayou) [inline src/pal/pal_open_enclave.h:21]:
```suggestion
    }
    
    static constexpr size_t address_bits = Aal::address_bits;
    static constexpr size_t page_size = Aal::smallest_page_size;
```

File: src/override/malloc-extensions.cc [hunk 3]
```diff
@@ -6,7 +6,8 @@ using namespace snmalloc;
 
 void get_malloc_info_v1(malloc_info_v1* stats)
 {
-  auto next_memory_usage = default_memory_provider().memory_usage();
-  stats->current_memory_usage = next_memory_usage.first;
-  stats->peak_memory_usage = next_memory_usage.second;
-}
\ No newline at end of file
+  auto unused_chunks = Globals::get_chunk_allocator_state().unused_memory();
+  auto peak = Globals::get_chunk_allocator_state().peak_memory_usage();
```
Comment (achamayou) [inline src/override/malloc-extensions.cc:10]:
```suggestion
  auto unused_chunks = Globals::get_chunk_allocator_state(nullptr).unused_memory();
  auto peak = Globals::get_chunk_allocator_state(nullptr).peak_memory_usage();
```

File: src/pal/pal_open_enclave.h (hunk 2 above)
Comment (mjp41) [reply src/pal/pal_open_enclave.h:21]: Resolved by #418

File: src/override/malloc-extensions.cc (hunk 3 above)
Comment (mjp41) [reply src/override/malloc-extensions.cc:10]: Resolved by #418

**Comment by mjp41** [comment]
I have now made the `snmalloc2` branch `main`, and the `master` branch `snmalloc1`.

**Pull request closed without merging**